File: 2018 ICPC Vietnam National Programming Contest/howtopaint.cs
using System;
using System.Linq;

class Program {
    const int MOD = 100003;
    const int MAXN = 1005;
    
    static long numerator = 1, denominator = 1, baseVal = 0;
    static int m, n;
    
    static void UpdateHook(int[] w) {
        System.Diagnostics.Debug.Assert(w[0] > 0);
        for (int i = 1; i < w.Length; i++) {
            System.Diagnostics.Debug.Assert(w[i] <= w[i - 1]);
        }
        
        bool[,] filled = new bool[MAXN, MAXN];
        int[,] ups = new int[MAXN, MAXN];
        int[,] rights = new int[MAXN, MAXN];
        
        for (int i = 0; i < w.Length; i++) {
            for (int j = 0; j < w[i]; j++) {
                filled[i, j] = true;
            }
        }
        
        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (filled[i, j]) {
                    ups[i, j] = (i + 1 < m && filled[i + 1, j]) ? (1 + ups[i + 1, j]) : 1;
                    rights[i, j] = (j + 1 < n && filled[i, j + 1]) ? (1 + rights[i, j + 1]) : 1;
                    int hook = ups[i, j] + rights[i, j] - 1;
                    denominator = (denominator * hook) % MOD;
                }
            }
        }
    }
    
    static void Multiply(int x) {
        while (x % MOD == 0) {
            baseVal++;
            x /= MOD;
        }
        numerator = (numerator * x) % MOD;
    }
    
    static void Main() {
        string[] input = Console.ReadLine().Split();
        m = int.Parse(input[0]);
        n = int.Parse(input[1]);
        System.Diagnostics.Debug.Assert(0 < m && m <= 1000);
        System.Diagnostics.Debug.Assert(0 < n && n <= 1000);
        
        int[] a = new int[m];
        int[] b = new int[m];
        int p = 0, q = 0;
        
        input = Console.ReadLine().Split();
        for (int i = 0; i < m; i++) {
            a[i] = int.Parse(input[i]);
            System.Diagnostics.Debug.Assert(a[i] <= n);
            b[i] = n - a[i];
            p += a[i];
            q += b[i];
        }
        
        Array.Sort(b);
        Array.Reverse(b);
        
        UpdateHook(a);
        UpdateHook(b);
        
        for (int i = 1; i <= p + q - 2; i++) {
            Multiply(i);
        }
        Multiply(p);
        Multiply(q);
        
        for (int i = 0; i < MOD - 2; i++) {
            numerator = (numerator * denominator) % MOD;
        }
        
        Console.WriteLine($"{baseVal} {numerator}");
    }
}

File: 2018 ICPC Vietnam National Programming Contest/howtopaint.py
MOD = 100003
MAXN = 1005

def main():
    import sys
    data = sys.stdin.read().split()
    
    m = int(data[0])
    n = int(data[1])
    assert 0 < m <= 1000
    assert 0 < n <= 1000
    
    a = list(map(int, data[2:2+m]))
    b = [n - x for x in a]
    
    p = sum(a)
    q = sum(b)
    
    b.sort(reverse=True)
    
    numerator = 1
    denominator = 1
    base = 0
    
    def update_hook(w):
        nonlocal denominator
        assert w[0] > 0
        for i in range(1, len(w)):
            assert w[i] <= w[i-1]
        
        filled = [[False] * MAXN for _ in range(MAXN)]
        ups = [[0] * MAXN for _ in range(MAXN)]
        rights = [[0] * MAXN for _ in range(MAXN)]
        
        for i in range(len(w)):
            for j in range(w[i]):
                filled[i][j] = True
        
        for i in range(m-1, -1, -1):
            for j in range(n-1, -1, -1):
                if filled[i][j]:
                    ups[i][j] = 1 + ups[i+1][j] if i+1 < m and filled[i+1][j] else 1
                    rights[i][j] = 1 + rights[i][j+1] if j+1 < n and filled[i][j+1] else 1
                    hook = ups[i][j] + rights[i][j] - 1
                    denominator = (denominator * hook) % MOD
    
    def multiply(x):
        nonlocal numerator, base
        while x % MOD == 0:
            base += 1
            x //= MOD
        numerator = (numerator * x) % MOD
    
    update_hook(a)
    update_hook(b)
    
    for i in range(1, p + q - 1):
        multiply(i)
    multiply(p)
    multiply(q)
    
    for i in range(MOD - 2):
        numerator = (numerator * denominator) % MOD
    
    print(f"{base} {numerator}")

if __name__ == "__main__":
    main()